### PR TITLE
CLN: Verify non-ndarray is an ExtensionArray in hash_array GH42003

### LINKED ISSUE #42003: BUG: 1.3.0rc1 pd.util.hash_array on Index fails to access _values_for_factorize
- [X] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas. (does not exist in 1.2.4, only seen in 1.3.0rc1 and master)

- [X] (optional) I have confirmed this bug exists on the master branch of pandas. (confirmed on 0b68d87a443)

---

#### Code Sample, a copy-pastable example

```python
In [1]: import pandas as pd

In [2]: pd.util.hash_array(pd.DatetimeIndex(['2018-10-28 01:20:00'], tz='Europe/Berlin'))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-389898c5af02> in <module>
----> 1 pd.util.hash_array(pd.DatetimeIndex(['2018-10-28 01:20:00'], tz='Europe/Berlin'))

~/working_dir/pandas/pandas/core/util/hashing.py in hash_array(vals, encoding, hash_key, categorize)
    287     elif not isinstance(vals, np.ndarray):
    288         # i.e. ExtensionArray
--> 289         vals, _ = vals._values_for_factorize()
    290 
    291     return _hash_ndarray(vals, encoding, hash_key, categorize)

AttributeError: 'DatetimeIndex' object has no attribute '_values_for_factorize'

In [3]: pd.util.hash_array(pd.Index([1,2,3]))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-e66efa244441> in <module>
----> 1 pd.util.hash_array(pd.Index([1,2,3]))

~/working_dir/pandas/pandas/core/util/hashing.py in hash_array(vals, encoding, hash_key, categorize)
    287     elif not isinstance(vals, np.ndarray):
    288         # i.e. ExtensionArray
--> 289         vals, _ = vals._values_for_factorize()
    290 
    291     return _hash_ndarray(vals, encoding, hash_key, categorize)

AttributeError: 'Int64Index' object has no attribute '_values_for_factorize'

In [4]: pd.util.hash_array(pd.RangeIndex(1,3))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-af0900aae979> in <module>
----> 1 pd.util.hash_array(pd.RangeIndex(1,3))

~/working_dir/pandas/pandas/core/util/hashing.py in hash_array(vals, encoding, hash_key, categorize)
    287     elif not isinstance(vals, np.ndarray):
    288         # i.e. ExtensionArray
--> 289         vals, _ = vals._values_for_factorize()
    290 
    291     return _hash_ndarray(vals, encoding, hash_key, categorize)

AttributeError: 'RangeIndex' object has no attribute '_values_for_factorize'
```

#### Problem description

This issue looks similar to https://github.com/pandas-dev/pandas/issues/41817, but that is specifically for DateTimeIndex with tz defined, while this seems to happen for any Index instance.

#### Expected Output

A hash of the input index, as in pandas < 1.3.0.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS      
------------------      
commit           : 2dd9e9ba9009a40191c0c0b96262fa3939d609f0
python           : 3.8.6.final.0
python-bits      : 64   
OS               : Linux 
OS-release       : 5.10.28-1rodete1-amd64
Version          : #1 SMP Debian 5.10.28-1rodete1 (2021-04-30)
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0rc1
numpy            : 1.19.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.2.1
setuptools       : 49.2.1
Cython           : 0.29.22
pytest           : 6.2.2
hypothesis       : 6.4.0
sphinx           : 3.5.1
blosc            : 1.10.2
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.21.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.7
fastparquet      : 0.5.0
gcsfs            : 0.7.2
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : 0.5.2
scipy            : 1.6.1
sqlalchemy       : 1.3.23
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.17.0
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.52.0
</details>

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Closes #42003

### CODE REVIEW

File: pandas/core/util/hashing.py [hunk 1]
```diff
@@ -286,6 +286,8 @@ def hash_array(
         return _hash_categorical(vals, encoding, hash_key)
     elif not isinstance(vals, np.ndarray):
```
Comment (jreback) [inline pandas/core/util/hashing.py:287]: i would rather explictily check for an extension array here. 

also pls add a test for this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/util/hashing.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/util/hashing.py:287]: There's already

```
    if not hasattr(vals, "dtype"):
        raise TypeError("must pass a ndarray-like")
```

at the top of the function for validation. Could maybe change that to check for EA/np.ndarray explicitly and change the message to be more user friendly.

Comment (jreback) [reply pandas/core/util/hashing.py:287]: yep agreed

Comment (TheNeuralBit) [reply pandas/core/util/hashing.py:287]: importing `ExtensionArray` so I can use it in an instance check is resulting in a circular import:
```
❯ python -m pytest pandas/tests/util/test_hashing.py 
ImportError while loading conftest '.../pandas/pandas/conftest.py'.
pandas/__init__.py:22: in <module>
    from pandas.compat import (
pandas/compat/__init__.py:15: in <module>
    from pandas.compat.numpy import (
pandas/compat/numpy/__init__.py:7: in <module>
    from pandas.util.version import Version
pandas/util/__init__.py:7: in <module>
    from pandas.core.util.hashing import (  # noqa
pandas/core/util/hashing.py:23: in <module>
    from pandas.core.arrays.base import ExtensionArray
pandas/core/arrays/__init__.py:1: in <module>
    from pandas.core.arrays.base import (
pandas/core/arrays/base.py:32: in <module>
    from pandas.compat import set_function_name
E   ImportError: cannot import name 'set_function_name' from partially initialized module 'pandas.compat' (most likely due to a circular import) (.../pandas/pandas/compat/__init__.py)
```

Any advice on how to resolve this? I could just move the import inside `hash_array`, but not sure if that's consistent with pandas coding conventions.

Comment (simonjayhawkins) [reply pandas/core/util/hashing.py:287]: we have a `is_array_like` in `pandas/core/dtypes/inference.py` that also includes `Index` and `Series` so is not suitable here.

Most public functions that accept an numpy array/EA probably accept a array-like even if not documented, like this one used to.

It is mainly internal functions where we would be more strict on the types accepted.

So we could maybe add a `is_array_like_strict` but we would normally already have validated the arguments that are passed internally so may not have much utility outside this public method which is more restrictive than most.

For now, I think OK to move the imports inside `hash_array` until we know we want to do this type of check elsewhere.

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears it has gone stale and needs to address the review. Closing but happy to reopen if you're still interested in continuing.

**Pull request closed without merging**